src/engine.rs

```diff
@@ -0,0 +1,2625 @@
+#[doc = include_str!("../docs/position_modeling.html")]
+#[doc = include_str!("../docs/stylus_state_modeler.html")]
+#[doc = include_str!("../docs/stroke_end.html")]
+impl StrokeModeler {
+    pub fn new(params: ModelerParams) -> Self {
+        Self {
+            params,
+            last_event: None,
+            last_corrected_event: None,
+            wobble_deque: VecDeque::with_capacity(
+                (2.0 * params.sampling_min_output_rate * params.wobble_smoother_timeout) as usize,
+            ),
+            wobble_duration_sum: 0.0,
+            wobble_weighted_pos_sum: (0.0, 0.0),
+            wobble_distance_sum: 0.0,
+            position_modeler: None,
+            state_modeler: StateModeler::new(params.stylus_state_modeler_max_input_samples),
+        }
+    }
+
+    /// Clears any in-progress stroke, keeping the same model parameters
+    /// Technically the error is not needed anymore as the params are kept
+    /// as is and can't be uninitialized
+    pub fn reset(&mut self) -> Result<(), i32> {
+        self.last_event = None;
+        self.wobble_deque = VecDeque::with_capacity(
+            (2.0 * self.params.sampling_min_output_rate * self.params.wobble_smoother_timeout)
+                as usize,
+        );
+        self.wobble_duration_sum = 0.0;
+        self.wobble_weighted_pos_sum = (0.0, 0.0);
+        self.position_modeler = None;
+        Ok(()) // to match
+    }
+
+    /// Clears any in-progress stroke, and re initialize the model with
+    /// the given parameters
+    ///
+    /// Here the error is also obsolete as the `ModelerParams` is expected
+    /// to have been built with [ModelerParams::validate] that validates the
+    /// parameters
+    pub fn reset_w_params(&mut self, params: ModelerParams) -> Result<(), i32> {
+        match params.validate() {
+            Ok(params) => self.params = params,
+            Err(_) => return Err(0),
+        };
+        self.last_event = None;
+        self.wobble_deque = VecDeque::with_capacity(
+            (2.0 * self.params.sampling_min_output_rate * self.params.wobble_smoother_timeout)
+                as usize,
+        );
+        self.wobble_duration_sum = 0.0;
+        self.wobble_weighted_pos_sum = (0.0, 0.0);
+        self.wobble_distance_sum = 0.0;
+        self.position_modeler = None;
+        Ok(())
+    }
+
+    /// Updates the model with a raw input, and appends newly generated Results to the results vector.
+    /// Any previously generated Result values remain valid.
+    /// (This does not require that any previous results returned remain in the results vector, as it is
+    /// appended to without examining the existing contents)
+    ///
+    /// If this does not return an error, results will contain at least one Result, and potentially
+    /// more if the inputs are slower than the minimum output rate
+    ///
+    /// for now rnote's wrapper codes verify that the input is not duplicated and time increases between strokes
```

Comment:
We should probably move the validation into ink-stroke-modeler-rs and remove the mention of Rnote, making it easier to use the the library for other projects